src/server/pachyderm_test.go

```diff
@@ -9653,6 +9653,90 @@ func TestListDatum(t *testing.T) {
 	require.Equal(t, 25, len(dis))
 }
 
+func TestListDatumFilter(t *testing.T) {
+	t.Parallel()
+	var (
+		ctx   = context.Background()
+		c, _  = minikubetestenv.AcquireCluster(t)
+		repo1 = tu.UniqueString("TestListDatum1")
+		repo2 = tu.UniqueString("TestListDatum2")
+	)
+
+	require.NoError(t, c.CreateRepo(repo1))
+	require.NoError(t, c.CreateRepo(repo2))
+
+	numFiles := 5
+	for i := 0; i < numFiles; i++ {
+		require.NoError(t, c.PutFile(client.NewCommit(repo1, "master", ""), fmt.Sprintf("file-%d", i), strings.NewReader("foo"), client.WithAppendPutFile()))
+		require.NoError(t, c.PutFile(client.NewCommit(repo2, "master", ""), fmt.Sprintf("file-%d", i), strings.NewReader("foo"), client.WithAppendPutFile()))
+	}
+
+	// filtering for failed should yield zero datums
+	s, err := c.PpsAPIClient.ListDatum(ctx, &pps.ListDatumRequest{
+		Filter: &pps.ListDatumRequest_Filter{State: []pps.DatumState{pps.DatumState_FAILED}},
+		Input: &pps.Input{
+			Cross: []*pps.Input{{
+				Pfs: &pps.PFSInput{
+					Repo: repo1,
+					Glob: "/*",
+				},
+			}, {
+				Pfs: &pps.PFSInput{
+					Repo: repo2,
+					Glob: "/*",
+				},
+			}},
+		},
+	})
+	require.NoError(t, err)
+
+	var i int
```

Comment:
We should probably add a common utility function for this to the clientsdk package. You can use the utility functions for the other streaming rpcs as reference implementations.

Reply:
I added a utility function to do this, and switched over to use it.

I am concerned that this kind of callback-based function is less idiomatic and immediately grokkable than just cycling through the client stream. It’s the sort of thing would would be completely the right thing to do in another language (e.g. in a Lisp macro), but feels like it’s not a great fit for Go.

Like I wrote though, I switched over as requested.

Reply:
Callback-based iteration is a pretty standard iteration pattern in Go (i.e. https://pkg.go.dev/path/filepath#Walk). Having the wrappers makes it simpler imo by removing boilerplate. It also gives us an opportunity to handle errors in a standard way.